concordium-contracts-common/src/types.rs

```diff
@@ -1081,25 +1081,149 @@ pub struct Cursor<T> {
     pub data:   T,
 }
 
+#[cfg(feature = "std")]
+impl std::error::Error for NewAttributeValueError {}
+
+/// Errors that can occur when constructing a new [`AttributeValue`].
+#[derive(Debug, PartialEq, Eq)]
+pub enum NewAttributeValueError {
+    TooLong(usize),
+}
+
+impl fmt::Display for NewAttributeValueError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            NewAttributeValueError::TooLong(size) => write!(
+                f,
+                "Attribute values have a max length of 31. The slice given had length {}.",
+                size
+            ),
+        }
+    }
+}
+
 /// Tag of an attribute. See the module [attributes](./attributes/index.html)
 /// for the currently supported attributes.
 #[repr(transparent)]
 #[cfg_attr(feature = "fuzz", derive(Arbitrary))]
 #[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq)]
 pub struct AttributeTag(pub u8);
 
-/// A borrowed attribute value. The slice will have at most 31 bytes.
+/// An attribute value.
 /// The meaning of the bytes is dependent on the type of the attribute.
-pub type AttributeValue<'a> = &'a [u8];
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct AttributeValue {
+    pub(crate) inner: [u8; 32],
+}
+
+impl AttributeValue {
+    /// Create a new [`Self`] from a slice of bytes. The slice must have a
+    /// length of *at most 31 bytes*.
+    pub fn new(data: &[u8]) -> Result<Self, NewAttributeValueError> {
```

Comment:
Minor thing, I find it a bit inconsistent to use a slice here, when `new_unchecked` does not.
I would prefer if we had a `new(data, [u8;32])` and a `try_from<&[u8]>`

Reply:
I decided to keep the `new` method because:

1. It is the easiest one to find for users.
2. The slice allows you to provide just the bytes you need instead of an array with padded zeroes (and a length).
3. We can't check that the length encoded in the first byte is correct (the sequence might end with one or more zeroes).

Instead, I added a `TryFrom` instance, which also makes sense to have :blush:.